src/coreclr/src/jit/lower.cpp

```diff
@@ -2986,6 +3017,175 @@ void Lowering::LowerRet(GenTreeUnOp* ret)
+            ret->gtOp1           = bitcast;
+            BlockRange().InsertBefore(ret, bitcast);
+            ContainCheckBitCast(bitcast);
+            break;
+        }
+        default:
+            unreached();
+    }
+}
+
+//----------------------------------------------------------------------------------------------
+// LowerRetStructLclVar: Lowers a return node with a struct lclVar as a source.
+//
+// Arguments:
+//     node - The return node to lower.
+//
+void Lowering::LowerRetStructLclVar(GenTreeUnOp* ret)
+{
+    assert(!comp->compDoOldStructRetyping());
+    assert(ret->OperIs(GT_RETURN));
+    GenTreeLclVar* lclVar = ret->gtGetOp1()->AsLclVar();
+    unsigned       lclNum = lclVar->GetLclNum();
+    LclVarDsc*     varDsc = comp->lvaGetDesc(lclNum);
+
+#ifdef DEBUG
+    if (comp->gtGetStructHandleIfPresent(lclVar) == NO_CLASS_HANDLE)
+    {
+        // a promoted struct field was retyped as its only field.
+        assert(varDsc->lvIsStructField);
+    }
+#endif
+    if (varDsc->lvPromoted && (comp->lvaGetPromotionType(lclNum) == Compiler::PROMOTION_TYPE_INDEPENDENT))
+    {
+        bool canEnregister = false;
+        if (varDsc->lvFieldCnt == 1)
+        {
+            // We have to replace it with its field.
+            assert(varDsc->lvRefCnt() == 0);
+            unsigned   fieldLclNum = varDsc->lvFieldLclStart;
+            LclVarDsc* fieldDsc    = comp->lvaGetDesc(fieldLclNum);
+            if (fieldDsc->lvFldOffset == 0)
+            {
+                lclVar->SetLclNum(fieldLclNum);
+                JITDUMP("Replacing an independently promoted local var with its only field for the return %u, %u\n",
+                        lclNum, fieldLclNum);
+                lclVar->ChangeType(fieldDsc->lvType);
+                canEnregister = true;
+            }
+        }
+        if (!canEnregister)
+        {
+            // TODO-1stClassStructs: We can no longer promote or enregister this struct,
+            // since it is referenced as a whole.
+            comp->lvaSetVarDoNotEnregister(lclNum DEBUGARG(Compiler::DNER_VMNeedsStackAddr));
+        }
+    }
+
+    var_types regType = varDsc->GetRegisterType(lclVar);
+    assert((regType != TYP_STRUCT) && (regType != TYP_UNKNOWN));
+    // Note: regType could be a small type, in this case return will generate an extension move.
+    lclVar->ChangeType(regType);
+    if (varTypeUsesFloatReg(ret) != varTypeUsesFloatReg(lclVar))
+    {
+        GenTreeUnOp* bitcast = new (comp, GT_BITCAST) GenTreeOp(GT_BITCAST, ret->TypeGet(), lclVar, nullptr);
+        ret->gtOp1           = bitcast;
+        BlockRange().InsertBefore(ret, bitcast);
+        ContainCheckBitCast(bitcast);
+    }
+}
+
+//----------------------------------------------------------------------------------------------
+// LowerCallStruct: Lowers a call node that returns a stuct.
+//
+// Arguments:
+//     call - The call node to lower.
+//
+// Note: it transforms the call's user.
```

Comment:
It is a bit unfortunate that we have to transform the call's user. I wonder whether it would be feasible and reasonable to handle these when lowering the user. I think this is fine for now.

Reply:
You are right, we have agreed to do that in the user lowering but when I did so I saw that:
1) we have to process call in `LowerCall` when it is unused, so compilation time is spent to find a user in any case;
2) the processing goes into 2 functions (in addition to `LowerCall`) `LowerBlockStore, LowerStoreIndir` and they are platform-specific, so the logic to find these cases should be duplicated for different platforms;
3) it takes more code and operations to find these cases from the user, because each struct call has its user that should be modified, but not all struct stores have such calls as they src.

so after a while, I changed it to process it in one place, I think it is more visible.